### DISCUSSION TITLE
Why dependency of LINKER_GENERIC_SECTIONS_PRESENT_AT_BOOT config is set in incorrect place?

### DISCUSSION THREAD

**Discussion by xakep-amatop** [zephyrproject-rtos/zephyr#56802] [Q&A] [answered]
As far as I know, it is not possible to use a condition in the same line where the boolean option is, so have we a mistake inside 
LINKER_GENERIC_SECTIONS_PRESENT_AT_BOOT config?
```
config LINKER_GENERIC_SECTIONS_PRESENT_AT_BOOT
	bool "Generic sections are present at boot" if DEMAND_PAGING && LINKER_USE_PINNED_SECTION
	default y
	help
	  When disabled, the linker sections other than the boot and
	  pinned sections will be marked as not present in the page
	  tables. This allows kernel to pull in data pages on demand
	  as required by current execution context when demand paging
	  is enabled. There is no need to load all code and data into
	  memory at once.

	  If unsure, say Y.
```

Best regards,
Mykola

**Comment by yashi** [comment]
Kconfig allows you to place a condition on a menu entry.  And there are many usage in Linux kernel.

> As far as I know, it is not possible to use a condition in the same line where the boolean option is

Do you have any reference for it?

[The doc](https://www.kernel.org/doc/html/latest/kbuild/kconfig-language.html#menu-attributes) doesn't explicitly state that you can do

```
<type> <prompt> <string> <cond>
```

But, because 

```
<type> <string>
```
is equivalent to
```
<type>
prompt <string>
```
and it explicity allows to do
```
prompt <string> <cond>
```
you can do
```
<type>
prompt <string> <cond>
```
which, you can reduced to
```
<type> <string> <cond>
```

The syntax might not allow this reduction. But, the a few lines down in the doc, you can find an example using a condition on `bool`:

> - dependencies: “depends on” <expr>
> 
>   This defines a dependency for this menu entry. If multiple dependencies are defined, they are connected with ‘&&’. Dependencies are applied to all other options within this menu entry (which also accept an “if” expression), so these two examples are equivalent:
> 
>   ```
>   bool "foo" if BAR
>   default y if BAR
>   ```
>   and:
> 
>   ```
>   depends on BAR
>   bool "foo"
>   default y
>   ```

Please refert to [the doc](https://www.kernel.org/doc/html/latest/kbuild/kconfig-language.html#menu-attributes) for more info.

**Comment by xakep-amatop** [comment]
Hi, @yashi 

it is quite interesting why I have different results with and without the next patch:
```
diff --git a/Kconfig.zephyr b/Kconfig.zephyr
index 2dd8492f75..6308b01977 100644
--- a/Kconfig.zephyr
+++ b/Kconfig.zephyr
@@ -254,7 +254,8 @@ config LINKER_USE_PINNED_SECTION
          board or custom linker script.
 
 config LINKER_GENERIC_SECTIONS_PRESENT_AT_BOOT
-       bool "Generic sections are present at boot" if DEMAND_PAGING && LINKER_USE_PINNED_SECTION
+       bool "Generic sections are present at boot"
+       depends on DEMAND_PAGING && LINKER_USE_PINNED_SECTION
        default y
        help
          When disabled, the linker sections other than the boot and
```

I built `samples/synchronization` two times with and without mentioned above patch and get different resulting app .confing-s:
```
index d187174ed2..0ed7f1c487 100644
--- a/build-spider-little-fs/zephyr/.config
+++ b/home/mykola_kvach/Documents/cfg1
@@ -906,6 +906,7 @@ CONFIG_SRAM_OFFSET=0
 #
 # CONFIG_LINKER_USE_BOOT_SECTION is not set
 # CONFIG_LINKER_USE_PINNED_SECTION is not set
+CONFIG_LINKER_GENERIC_SECTIONS_PRESENT_AT_BOOT=y
 CONFIG_LINKER_LAST_SECTION_ID=y
 CONFIG_LINKER_LAST_SECTION_ID_PATTERN=0xE015E015
 CONFIG_LINKER_USE_RELAX=y
```

Did I miss something ?

**Reply by yashi** [reply, marked as answer]
Ah, that's different.

The `if` on the `bool` line is to show or hide the prompt.  It doesn't have anything to do with the default value.

If you want to control the default value depending on the dependencies, you must have `if` on the `default` line.  Just like the quote I placed on the other answer.

```
bool "foo" if BAR
default y if BAR
```

Thus, you need the following patch if that's what it intends to do.

```diff
diff --git a/Kconfig.zephyr b/Kconfig.zephyr
index 8fa4faec87c..47a5450c458 100644
--- a/Kconfig.zephyr
+++ b/Kconfig.zephyr
@@ -248,7 +248,7 @@ config LINKER_USE_PINNED_SECTION
 
 config LINKER_GENERIC_SECTIONS_PRESENT_AT_BOOT
 	bool "Generic sections are present at boot" if DEMAND_PAGING && LINKER_USE_PINNED_SECTION
-	default y
+	default y if DEMAND_PAGING && LINKER_USE_PINNED_SECTION
 	help
 	  When disabled, the linker sections other than the boot and
 	  pinned sections will be marked as not present in the page
```

**Reply by yashi** [reply]
I haven't check how `LINKER_GENERIC_SECTIONS_PRESENT_AT_BOOT` is used.  If the intetion is as mentioned above, please open an issue.

Thanks.

**Reply by xakep-amatop** [reply]
@yashi, thank you
